### PR TITLE
refactor: use typed Trace

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

closes: #6362 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved error messages for trace validation to provide clearer guidance.

- **Refactor**
  - Updated internal data structures for denomination tracing to enhance consistency and reliability.

- **New Features**
  - Added new test cases to cover a wider range of scenarios for token and packet validation.

- **Documentation**
  - Adjusted deprecated options formatting in the `Query` service for better readability.

### CODE REVIEW

File: modules/apps/transfer/types/packet_test.go [hunk 1]
```diff
@@ -27,9 +27,18 @@ func TestFungibleTokenPacketDataValidateBasic(t *testing.T) {
 		expPass    bool
 	}{
 		{"valid packet", types.NewFungibleTokenPacketData(denom, amount, sender, receiver, ""), true},
+		{"valid packet, base denom with leading slash", types.NewFungibleTokenPacketData("transfer/channel-1/uatom/", amount, sender, receiver, ""), true},
```
Comment (colin-axner) [inline modules/apps/transfer/types/packet_test.go:30]: pulled from `TestValidatePrefixedDenom`

File: modules/apps/transfer/types/trace_test.go [hunk 2]
```diff
@@ -94,37 +94,6 @@ func TestDenomTrace_Validate(t *testing.T) {
 	}
 }
 
-func TestValidatePrefixedDenom(t *testing.T) {
```
Comment (colin-axner) [inline modules/apps/transfer/types/trace_test.go:97]: moved relevant test cases to packet data validate test func, but maybe can be moved to the extract test instead

File: modules/apps/transfer/types/trace.go [hunk 3]
```diff
@@ -29,8 +58,10 @@ import (
 // - "uatom" => DenomTrace{Path: "", BaseDenom: "uatom"}
 func ParseDenomTrace(rawDenom string) DenomTrace {
 	denom := ExtractDenomFromFullPath(rawDenom)
+	path := denom.FullPath()
+	path = strings.TrimSuffix(path, denom.Base)
```
Comment (colin-axner) [inline modules/apps/transfer/types/trace.go:62]: quick hack for now, I plan to remove this function in the next pr

File: modules/apps/transfer/types/trace.go [hunk 4]
```diff
@@ -17,6 +17,35 @@ import (
...
+
+// String returns the Trace the format:
+// <portID>/<channelID>
+func (t Trace) String() string {
+	var sb strings.Builder
+	sb.WriteString(t.PortId)    // nolint:revive // no error returned by WriteString
+	sb.WriteByte('/')           //nolint:revive // no error returned by WriteByte
+	sb.WriteString(t.ChannelId) // nolint:revive
+	return sb.String()
+}
```
Comment (chatton) [inline modules/apps/transfer/types/trace.go:47]: when there's no logic involved, should we just do a direct return `fmt.Sprintf("%s/%s", t.PortId, t.ChannelId)`?

**Review by chatton — APPROVED** [review]
LGTM pending the linter fixes, had one suggestions but nothing major, nice clean refactor!

File: modules/apps/transfer/types/trace.go (hunk 4 above)
Comment (colin-axner) [reply modules/apps/transfer/types/trace.go:47]: that sounds good to me :+1:

File: modules/apps/transfer/ibc_module_test.go [hunk 5]
```diff
@@ -605,7 +605,7 @@ func (suite *TransferTestSuite) TestPacketDataUnmarshalerInterface() {
 						{
 							Denom: types.Denom{
 								Base:  ibctesting.TestCoin.Denom,
-								Trace: []types.Trace{types.Trace{}},
+								Trace: []types.Trace{{}},
```
Comment (colin-axner) [inline modules/apps/transfer/ibc_module_test.go:608]: I find this to be an odd linting standard

**Pull request merged**